core/src/main/java/org/elasticsearch/common/lucene/index/ESIndexDeletionPolicy.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.common.lucene.index;
+
+import org.apache.lucene.index.IndexCommit;
+import org.apache.lucene.index.IndexDeletionPolicy;
+
+import java.io.IOException;
+import java.util.List;
+
+/**
+ * An {@link IndexDeletionPolicy} that deletes unneeded index commits, and returns index commits are not deleted by this policy.
+ */
+public interface ESIndexDeletionPolicy {
+
+    /**
+     * Similar to {@link IndexDeletionPolicy#onInit(List)} but returns a list of kept index commits.
+     *
+     * @param commits A list of index commits sorted by age (the 0th one is the oldest commit).
+     * @return a list of index commits that are not deleted by this policy.
+     */
+    List<IndexCommit> onInit(List<? extends IndexCommit> commits) throws IOException;
+
+    /**
+     * Similar to {@link IndexDeletionPolicy#onCommit(List)} but returns a list of kept index commits.
+     *
+     * @param commits A list of index commits sorted by age (the 0th one is the oldest commit).
+     * @return a list of index commits that are not deleted by this policy.
+     */
+    List<IndexCommit> onCommit(List<? extends IndexCommit> commits) throws IOException;
+}
```

Comment:
This interface clearly came from `IndexDeletionPolicy` but I think it's now different enough that it's worth breaking the link entirely and coming up with more descriptive names. The `onInit()` and `onCommit()` methods of `IndexDeletionPolicy` return `void` but here we're returning a list of the kept commits. Additionally, I think all the implementations just delegate `onInit()` to `onCommit()` (except `KeepUntilGlobalCheckpointDeletionPolicy` that special-cases an empty argument).